### PR TITLE
Fix(Android): Emulator boot not detected >= api 31

### PR DESCRIPTION
**Pull request by jonathanmos**
## Description
Detect when emulator boots >= api 31

Fixes: https://github.com/wix/Detox/issues/3071
Fixes: https://github.com/wix/Detox/issues/3342

### CODE REVIEW

**Review by noomorph — CHANGES_REQUESTED** [review]
@jonathanmos think about a different approach: https://github.com/wix/Detox/compare/next...draft/android-emulator?expand=1

File: detox/src/devices/allocation/drivers/android/emulator/launchEmulatorProcess.js [hunk 1]
```diff
@@ -1,29 +1,15 @@
 const fs = require('fs');
 
 const _ = require('lodash');
-const { Tail } = require('tail');
```
Comment (d4vidi) [inline detox/src/devices/allocation/drivers/android/emulator/launchEmulatorProcess.js:4]: Does this mean we can get rid of `tail` as a module altogether?

File: detox/src/devices/common/drivers/android/exec/ADB.js [hunk 2]
```diff
@@ -189,6 +189,10 @@ class ADB {
     }
   }
 
+  async waitForDevice(deviceId) {
+    return await this.adbCmd(deviceId, 'wait-for-device');
```
Comment (d4vidi) [inline detox/src/devices/common/drivers/android/exec/ADB.js:193]: Test coverage

File: detox/src/devices/allocation/drivers/android/emulator/launchEmulatorProcess.js [hunk 3]
```diff
@@ -46,6 +31,8 @@ function launchEmulatorProcess(emulatorName, emulatorExec, emulatorLaunchCommand
 
   log = log.child({ child_pid: childProcessPromise.childProcess.pid });
 
+  adb.waitForDevice(adbName).then(() => childProcessPromise._cpResolve());
```
Comment (d4vidi) [inline detox/src/devices/allocation/drivers/android/emulator/launchEmulatorProcess.js:34]: What happens if something indeed goes wrong here? Emulator crash or whatever? Will it throw or hang forever?

**Review by d4vidi — COMMENTED** [review]
Generally approved with these 2 caveats:
- In the very least, this has to be tested manually against numerous SDK's, cold/quick boots etc. using new / old adb binaries and make sure wait-for-boot always applies.
- Need to set up an automated testing environment for this, in order to test SDK ≥31 regularly (at least the launch phase).

File: detox/src/devices/common/drivers/android/exec/ADB.js (hunk 2 above)
Comment (jonathanmos) [reply detox/src/devices/common/drivers/android/exec/ADB.js:193]: Added unit test for this method

**Review by d4vidi — APPROVED** [review]

**Review by noomorph — APPROVED** [review]

**Pull request merged**